File: IO_HIL/FDX/EasyClient/FDXDatagram.cpp
/*-----------------------------------------------------------------------------
Module: CANoeFDX
Interfaces: FDXDatagram
-------------------------------------------------------------------------------
Datagram builder class for CANoe FDX protocol
-------------------------------------------------------------------------------
Copyright (c) Vector Informatik GmbH. All rights reserved.
-----------------------------------------------------------------------------*/


#include "stdafx.h"
#include "FDXDatagram.h"
#include "assert.h"

using namespace CANoeFDX;

FDXDatagram::FDXDatagram()
: mSize(0)
{
  memset(mBuffer, 0, cCapacity);
}



void FDXDatagram::InitWithHeader()
{
  memset(mBuffer, 0, cCapacity);
  CANoeFDX::DatagramHeader* header = reinterpret_cast<CANoeFDX::DatagramHeader*>(mBuffer);
  header->fdxSignature = CANoeFDX::kFdxSignature;
  header->fdxMajorVersion = CANoeFDX::kFdxMajorVersion;
  header->fdxMinorVersion = CANoeFDX::kFdxMinorVersion;
  header->numberOfCommands = 0;
  header->sequenceNumber = CANoeFDX::kSequenceNumberUnused;
  header->fdxProtocolFlags = kByteOrderLittleEndian;
  header->reserved = 0;
  mSize = sizeof(CANoeFDX::DatagramHeader);
}


void FDXDatagram::SetSequenceNumber(CANoeFDX::uint16 sequenceNumber)
{
  CANoeFDX::DatagramHeader* header = reinterpret_cast<CANoeFDX::DatagramHeader*>(mBuffer);
  header->sequenceNumber = sequenceNumber;
}


void FDXDatagram::AddDataRequest(CANoeFDX::uint16 groupID)
{
  CANoeFDX::DataRequestCommand* command = static_cast<CANoeFDX::DataRequestCommand*>(AddCommand(CANoeFDX::kCommandCode_DataRequest, sizeof(CANoeFDX::DataRequestCommand)));
  command->groupID = groupID;
}



void* FDXDatagram::AddDataExchange(CANoeFDX::uint16 groupID, CANoeFDX::uint16 dataSize)
{
  CANoeFDX::DataExchangeCommand* command = static_cast<CANoeFDX::DataExchangeCommand*>(AddCommand(CANoeFDX::kCommandCode_DataExchange, CANoeFDX::kDataExchangeBaseSize + dataSize));
  command->groupID = groupID;
  command->dataSize = dataSize;
  return command->dataBytes;
}



void FDXDatagram::AddFreeRunningRequest(CANoeFDX::uint16 groupID, CANoeFDX::uint16 flags, CANoeFDX::uint32 cycleTime, CANoeFDX::uint32 firstDuration)
{
  CANoeFDX::FreeRunningRequestCommand* command = static_cast<CANoeFDX::FreeRunningRequestCommand*>(AddCommand(CANoeFDX::kCommandCode_FreeRunningRequest, sizeof(CANoeFDX::FreeRunningRequestCommand)));
  command->groupID = groupID;
  command->flags = flags;
  command->cycleTime = cycleTime;
  command->firstDuration = firstDuration;
}


void FDXDatagram::AddFreeRunningCancel(CANoeFDX::uint16 groupID)
{
  CANoeFDX::FreeRunningCancelCommand* command = static_cast<CANoeFDX::FreeRunningCancelCommand*>(AddCommand(CANoeFDX::kCommandCode_FreeRunningCancel, sizeof(CANoeFDX::FreeRunningCancelCommand)));
  command->groupID = groupID;
}



void FDXDatagram::AddStart()
{
  AddCommand(CANoeFDX::kCommandCode_Start, sizeof(CANoeFDX::CommandHeader));
}



void FDXDatagram::AddStop()
{
  AddCommand(CANoeFDX::kCommandCode_Stop, sizeof(CANoeFDX::CommandHeader));
}



CANoeFDX::CommandHeader* FDXDatagram::AddCommand(CANoeFDX::uint16 commandCode, CANoeFDX::uint16 commandSize)
{
  CANoeFDX::DatagramHeader* header  = reinterpret_cast<CANoeFDX::DatagramHeader*>(mBuffer);
  CANoeFDX::CommandHeader*  command = reinterpret_cast<CANoeFDX::CommandHeader*>(mBuffer+mSize);
  assert(mSize+commandSize <= cCapacity);
  mSize += commandSize;
  header->numberOfCommands++;
  command->commandSize = commandSize;
  command->commandCode = commandCode;
  return command;
}




CANoeFDX::uint16 IncrementSequenceNumber(CANoeFDX::uint16 seqNr)
{
  if (seqNr==CANoeFDX::kSequenceNumberUnused)
  {
    return CANoeFDX::kSequenceNumberUnused;
  }
  else if (seqNr&CANoeFDX::kSequenceNumberSessionEndFlag)
  {
    return CANoeFDX::kSequenceNumberUnused;
  }
  else if (seqNr==0x7FFF)
  {
    return 1;
  }
  else
  {
    return seqNr+1;
  }
}

File: A429/A429BasicFDX/FDXClient_C#/FDXDatagram.cs
/*-----------------------------------------------------------------------------
Module: CANoeFDX
Interfaces: FDXDatagram
-------------------------------------------------------------------------------
Datagram builder class for CANoe FDX protocol
-------------------------------------------------------------------------------
Copyright (c) Vector Informatik GmbH. All rights reserved.
-----------------------------------------------------------------------------*/
using System;
using System.Collections.Generic;

namespace FDXClient
{
  public class FDXDatagram
  {
    public FDXDatagram()
    {
      this.mHeader = new DatagramHeader();
      this.mCommands = new List<CommandHeader>();
    }

    public void InitWithHeader()
    {
      this.mHeader.FDXSignature = FDXHelper.kFDXSignature;
      this.mHeader.FDXMajorVersion = FDXHelper.kFDXMajorVersion;
      this.mHeader.FDXMinorVersion = FDXHelper.kFDXMinorVersion;
      this.mHeader.NumberOfCommands = 0;
      this.mHeader.SequenceNumber = FDXHelperSequenceNumber.kSequenceNumberUnused;
      this.mHeader.Reserved = 0;

      // the command list gets cleared when a new datagram is initialized
      this.mCommands.Clear();
      this.Size = 16;
    }

    public void SetSequenceNumber(ushort sequenceNumber)
    {
      this.mHeader.SequenceNumber = sequenceNumber;
    }

    public void AddDataRequest(ushort groupId)
    {
      this.AddCommand(new DataRequestCommand{ CommandCode = FDXHelperCommandCode.kCommandCodeDataRequest, CommandSize = 6, GroupId = groupId});
      this.Size += 6;
    }

    public void AddDataExchange(ushort groupId, byte[] data)
    {
      this.AddCommand(new DataExchangeCommand { CommandCode = FDXHelperCommandCode.kCommandCodeDataExchange, CommandSize = (ushort)(8 + data.Length), GroupId = groupId, DataSize = (ushort)data.Length, DataBytes = data });
      this.Size += (uint)(8 + data.Length);
    }

    public void AddStartCommand()
    {
      this.AddCommand(new CommandHeader { CommandSize = 4, CommandCode = FDXHelperCommandCode.kCommandCodeStart });
      this.Size += 4;
    }

    public void AddStopCommand()
    {
      this.AddCommand(new CommandHeader { CommandSize = 4, CommandCode = FDXHelperCommandCode.kCommandCodeStop });
      this.Size += 4;
    }

    public void AddStatusRequest()
    {
      this.AddCommand(new CommandHeader { CommandSize = 4, CommandCode = FDXHelperCommandCode.kCommandCodeStatusRequest});
      this.Size += 4;
    }

    public void AddKeyCommand(UInt32 keyCode)
    {
      this.AddCommand(new KeyCommand{CommandSize = 8, CommandCode = FDXHelperCommandCode.kCommandCodeKey, CanoeKeyCode = keyCode});
      this.Size += 8;
    }
    
    public byte[] GenerateBuffer()
    {
      this.mBuffer = new byte[this.Size];
      var arrHeader = this.mHeader.Serialize();
      var arrPos = 0;
      Array.Copy(arrHeader, 0, this.mBuffer, arrPos, arrHeader.Length);
      arrPos = arrHeader.Length;
      foreach (var command in this.mCommands)
      {
        var tmpArr = command.Serialize();
        Array.Copy(tmpArr, 0, this.mBuffer, arrPos, tmpArr.Length);
        arrPos += tmpArr.Length;
      }
      return this.mBuffer;
    }

    public byte[] Buffer
    {
      get { return this.mBuffer; }
      set { this.mBuffer = value; }
    }

    public static UInt16 IncrementSequenceNumber(UInt16 seqNr)
    {
      if (seqNr == FDXHelperSequenceNumber.kSequenceNumberUnused)
      {
        return FDXHelperSequenceNumber.kSequenceNumberUnused;
      }
      else if ((seqNr & FDXHelperSequenceNumber.kSequenceNumberSessionEndFlag) > 0)
      {
        return FDXHelperSequenceNumber.kSequenceNumberUnused;
      }
      else if (seqNr == 0x7FFF)
      {
        return 1;
      }
      else
      {
        return (ushort)(seqNr + 1);
      }
    }

    private void AddCommand(CommandHeader newCommand)
    {
      this.mCommands.Add(newCommand);
      this.mHeader.NumberOfCommands++;
    }

    #region Variables
    public uint Size { get; set; }
    private readonly DatagramHeader mHeader;
    private readonly List<CommandHeader> mCommands;
    private byte[] mBuffer;
    #endregion
  }
}
